### PR TITLE
add Networking API

### PR DESCRIPTION
**Pull request by SpaceWalkerRS**
Adds a Networking module and an implementation for (13w41a)-(1.13-pre2), with the following features:
- Login and disconnection events for both the client and server sides.
- APIs for sending and receiving data between the client and server.

## Networking API

Sending and receiving data is done through the `ClientPlayNetworking` and `ServerPlayNetworking` classes. Mods can register network listeners through the `registerListener` and `registerListenerAsync` methods, allowing them to receive data through specific channels. Sending data is done through the `send` methods.

Upon login, the client will initiate a "handshake" with the server, in which it will let the server know which channels the client has listeners for. The server will then reciprocate by telling the client which channels the server has listeners for. This information is then used to close channels for which there are no listeners. This is to avoid packet and log spam when one side has a mod installed but the other does not.

### CODE REVIEW

**Comment by SpaceWalkerRS** [comment]
We had some discussion before about whether the Networking API should send/receive data in NBT format. I decided against this for now and use the `PacketByteBuf` in order to keep the implementation as "raw" as possible. This still leaves the option open for mods utilize NBT for their networking, but encourages serializing to/from the `PacketByteBuf`directly, which is probably preferable if you're sending larger amounts of data.

**Review by SRAZKVT — COMMENTED** [review]
looks good, seems to be lacking some documentation overall though, but it looks alright to me

**Comment by SpaceWalkerRS** [comment]
> looks good, seems to be lacking some documentation overall though, but it looks alright to me

Good point. I'll go ahead and write some before merging this in.

File: libraries/networking/networking-mc13w41a-mc1.13-pre2/src/main/java/net/ornithemc/osl/networking/impl/CommonChannels.java [hunk 1]
```diff
@@ -0,0 +1,7 @@
+package net.ornithemc.osl.networking.impl;
+
+public final class CommonChannels {
+
+	public static final String CHANNELS = "osl|channels";
```
[the remark this answers is not in the document]
Comment (SpaceWalkerRS) [reply libraries/networking/networking-mc13w41a-mc1.13-pre2/src/main/java/net/ornithemc/osl/networking/impl/CommonChannels.java:12]: I added some javadocs to this, although it's not super relevant as it's in the `impl` package and for internal use only.

[the remark this answers is not in the document]
Comment (SRAZKVT) [reply libraries/networking/networking-mc13w41a-mc1.13-pre2/src/main/java/net/ornithemc/osl/networking/impl/CommonChannels.java:12]: should still imo explain what components of the string do what, that part currently seems pretty cryptic

**Review by SRAZKVT — COMMENTED** [review]
looks good

File: libraries/networking/networking-mc13w41a-mc1.13-pre2/src/main/java/net/ornithemc/osl/networking/impl/CommonChannels.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (SpaceWalkerRS) [reply libraries/networking/networking-mc13w41a-mc1.13-pre2/src/main/java/net/ornithemc/osl/networking/impl/CommonChannels.java:12]: Fair. It's just following the conventions of Vanilla custom payload packets mostly.

**Pull request merged**